src/PasswordPolicy.php

```diff
@@ -0,0 +1,308 @@
+use Drupal\user\Entity\Role;
+use Drupal\user\RoleInterface;
+
+/**
+ * Defines the Password Policy.
+ */
+final class PasswordPolicy implements PasswordPolicyInterface {
+
+  /**
+   * Expiry of the password in seconds, 0 means that passwords doesn't expire.
+   *
+   * @var int
+   */
+  private $expireSeconds;
+
+  /**
+   * Show warning message before the password would expire in a given seconds.
+   *
+   * @var int
+   */
+  private $expireWarnSeconds;
+
+  /**
+   * Expiry warning message.
+   *
+   * @var string|null
+   */
+  private $expiryWarningMessage;
+
+  /**
+   * The number of constraints that are required.
+   *
+   * @var int
+   */
+  private $minimumRequiredConstraints;
+
+  /**
+   * Role ID.
+   *
+   * @var string
+   */
+  private $role;
+
+  /**
+   * The array of password constraints for this password policy.
+   *
+   * @var array
+   */
+  private $constraints = [];
+
+  /**
+   * Holds the collection of password constraints that are used by this policy.
+   *
+   * @var \Drupal\password_enhancements\PasswordConstraintPluginCollection
+   */
+  private $constraintsCollection;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(RoleInterface $role, int $expire_seconds, int $expire_warn_seconds, string $expiry_warning_message, int $minimum_required_constraints, array $constraints) {
+    $this->role = $role;
+    // Set initial values.
+    $this->expireSeconds = $expire_seconds;
+    $this->expireWarnSeconds = $expire_warn_seconds;
+    $this->expiryWarningMessage = $expiry_warning_message;
+    $this->minimumRequiredConstraints = $minimum_required_constraints;
+    $this->constraints = $constraints;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function createFromRole(RoleInterface $role): ?PasswordPolicyInterface {
+    if ($third_party_settings = $role->getThirdPartySettings('password_enhancements')) {
+      return new static($role,
+        $third_party_settings['expireSeconds'] ?? self::PASSWORD_NO_EXPIRY,
+        $third_party_settings['expireWarnSeconds'] ?? self::PASSWORD_NO_WARNING,
+        $third_party_settings['expiryWarningMessage'] ?? 'Your password will expire on @date_time, please <a href="@url">change your password</a> before it expires to prevent any potential data loss.',
+        $third_party_settings['minimumRequiredConstraints'] ?? 1,
+        $third_party_settings['constraints'] ?? []);
+    }
+    return NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function loadByRoles(array $roles): ?PasswordPolicyInterface {
+    $policies = static::loadMultipleByRoles($roles);
+    return $policies ? reset($policies) : NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function loadMultipleByRoles(array $roles): array {
+    // Load all the roles asked for, but filter out the ones without policies.
+    $role_objects = array_filter(Role::loadMultiple($roles), function (RoleInterface $role) {
```

Comment:
IIRC the entities should be loaded through their entity storage and we shouldn't call the load function directly on them.

Reply:
Hmm. I thought the reasoning was fair, but it turns out that `\Drupal\Core\Entity\EntityBase::loadMultiple()` does use the entity storage – just like we'd do if we'd load the entity storage using `\Drupal::entityTypeManager()->getStorage()`. So I'm not seeing the reason for changing this after all.

Reply:
Because this is not a service and also a static method we cannot really DI the `entityTypeManager` service.